# Module 5 Homework - Cleaning Up the MoMA Artwork 

In this lab, you will put your Pandas data cleaning skills to the test by cleaning up the MoMA artwork csv file.  Solve each of the following problems using the functions from `dfply`.  

For each problem, perform two steps.

1. Explore the specified column using the techniques from the last lecture.
2. Fix all issues found in part 1.

**Note that I will be looking for a literate, easy to follow, set of code and markdown cells.**  Part of your score will be determined by the cleanliness/conciseness of your presentation.

Where possible, solve each problem with one pipe.

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pylab as plt
from dfply import *
import numpy as np
import re
from more_dfply import extract
from more_dfply import ifelse
from more_dfply import fix_names
%matplotlib inline

In [2]:
artwork = pd.read_csv("./data/Artworks.csv")

## Problem 1
    
**Task:** Explore the values of `artist.EndDate` and fix any issues.


In [3]:
artwork = (artwork >> fix_names)

Fixing the Names in the artwork data set by removing unwanted characters

In [4]:
artwork.head()

Title  \
0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1  City of Music, National Superior Conservatory ...   
2  Villa near Vienna Project, Outside Vienna, Aus...   
3  The Manhattan Transcripts Project, New York, N...   
4  Villa, project, outside Vienna, Austria, Exter...   

                     Artist ConstituentID  \
0               Otto Wagner          6210   
1  Christian de Portzamparc          7470   
2                Emil Hoppe          7605   
3           Bernard Tschumi          7056   
4                Emil Hoppe          7605   

                                   ArtistBio Nationality BeginDate EndDate  \
0                      (Austrian, 1841–1918)  (Austrian)    (1841)  (1918)   
1                        (French, born 1944)    (French)    (1944)     (0)   
2                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   
3  (French and Swiss, born Switzerland 1944)          ()    (1944)     (0)   
4                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   

   Gender  Date                                             Medium  ...  \
0  (Male)  1896      Ink and cut-and-pasted painted pages on paper  ...   
1  (Male)  1987                  Paint and colored pencil on print  ...   
2  (Male)  1903  Graphite, pen, color pencil, ink, and gouache ...  ...   
3  (Male)  1980  Photographic reproduction with colored synthet...  ...   
4  (Male)  1903  Graphite, color pencil, ink, and gouache on tr...  ...   

                                        ThumbnailURL Circumference_cm  \
0  http://www.moma.org/media/W1siZiIsIjU5NDA1Il0s...              NaN   
1  http://www.moma.org/media/W1siZiIsIjk3Il0sWyJw...              NaN   
2  http://www.moma.org/media/W1siZiIsIjk4Il0sWyJw...              NaN   
3  http://www.moma.org/media/W1siZiIsIjEyNCJdLFsi...              NaN   
4  http://www.moma.org/media/W1siZiIsIjEyNiJdLFsi...              NaN   

  Depth_cm Diameter_cm Height_cm Length_cm Weight_kg  Width_cm Seat_Height_cm  \
0      NaN         NaN   48.6000       NaN       NaN  168.9000            NaN   
1      NaN         NaN   40.6401       NaN       NaN   29.8451            NaN   
2      NaN         NaN   34.3000       NaN       NaN   31.8000            NaN   
3      NaN         NaN   50.8000       NaN       NaN   50.8000            NaN   
4      NaN         NaN   38.4000       NaN       NaN   19.1000            NaN   

  Duration_sec  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 29 columns]

Looking to see if the column names look nice after fixing them. Also exploring the data.

In [5]:
artwork.EndDate.head(10)

0    (1918)
1       (0)
2    (1957)
3       (0)
4    (1957)
5       (0)
6       (0)
7       (0)
8       (0)
9       (0)
Name: EndDate, dtype: object

Looking at the EndDate column. There needs to be cleaning as the dates a in ()

In [6]:
artwork = (artwork
>> mutate(EndDate_strip = X.EndDate.str.replace('[()]', ''))
)
artwork.head(5)

/home/vn6415dw/.pyenv/versions/anaconda3-2020.02/lib/python3.7/site-packages/dfply/base.py:82: FutureWarning: The default value of regex will change from True to False in a future version.
  **_context_kwargs(kwargs)(x)),


Title  \
0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1  City of Music, National Superior Conservatory ...   
2  Villa near Vienna Project, Outside Vienna, Aus...   
3  The Manhattan Transcripts Project, New York, N...   
4  Villa, project, outside Vienna, Austria, Exter...   

                     Artist ConstituentID  \
0               Otto Wagner          6210   
1  Christian de Portzamparc          7470   
2                Emil Hoppe          7605   
3           Bernard Tschumi          7056   
4                Emil Hoppe          7605   

                                   ArtistBio Nationality BeginDate EndDate  \
0                      (Austrian, 1841–1918)  (Austrian)    (1841)  (1918)   
1                        (French, born 1944)    (French)    (1944)     (0)   
2                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   
3  (French and Swiss, born Switzerland 1944)          ()    (1944)     (0)   
4                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   

   Gender  Date                                             Medium  ...  \
0  (Male)  1896      Ink and cut-and-pasted painted pages on paper  ...   
1  (Male)  1987                  Paint and colored pencil on print  ...   
2  (Male)  1903  Graphite, pen, color pencil, ink, and gouache ...  ...   
3  (Male)  1980  Photographic reproduction with colored synthet...  ...   
4  (Male)  1903  Graphite, color pencil, ink, and gouache on tr...  ...   

  Circumference_cm Depth_cm Diameter_cm Height_cm Length_cm Weight_kg  \
0              NaN      NaN         NaN   48.6000       NaN       NaN   
1              NaN      NaN         NaN   40.6401       NaN       NaN   
2              NaN      NaN         NaN   34.3000       NaN       NaN   
3              NaN      NaN         NaN   50.8000       NaN       NaN   
4              NaN      NaN         NaN   38.4000       NaN       NaN   

   Width_cm  Seat_Height_cm Duration_sec EndDate_strip  
0  168.9000             NaN          NaN          1918  
1   29.8451             NaN          NaN             0  
2   31.8000             NaN          NaN          1957  
3   50.8000             NaN          NaN             0  
4   19.1000             NaN          NaN          1957  

[5 rows x 30 columns]

Cleaning the EndDate column by stripping the parenthesis out of the names. So (1934) will be 1934. 

In [7]:
artwork.EndDate_strip.unique()

array(['1918', '0', '1957', ..., '0 1943 1966 1992 1975',
       '0 1943 1966 1975', '0 1934 1933'], dtype=object)

Looking at the unique values in the EndDate column. It looks like there are multiple entries in the same spot. '0 1934 1933' is a good example

In [8]:
artwork['EndDate_strip'] = artwork['EndDate_strip'].astype(str)

Making the EndDate_strip column into type string instead of object

In [9]:
max(artwork.EndDate_strip,key=len)

'0 1985 1979 1981 1965 1961 1989 1980 2002 1990 1956 2000 1973 1988 1985 0 2003 2013 1995 1971 1975 1994 1992 1981 1974 2002 1961 1943 1984 1997 1965'

To clean this using minimal code I need to find the length of the biggest row that has the most entries. There appear to be a max of 31 values in one row in EndDate. This should be even throughout cleaning the others like gender and BeginDate

In [10]:
EndDate_names = ['EndDate_{0}'.format(c) for c in range(31)]
EndDate_names

['EndDate_0',
 'EndDate_1',
 'EndDate_2',
 'EndDate_3',
 'EndDate_4',
 'EndDate_5',
 'EndDate_6',
 'EndDate_7',
 'EndDate_8',
 'EndDate_9',
 'EndDate_10',
 'EndDate_11',
 'EndDate_12',
 'EndDate_13',
 'EndDate_14',
 'EndDate_15',
 'EndDate_16',
 'EndDate_17',
 'EndDate_18',
 'EndDate_19',
 'EndDate_20',
 'EndDate_21',
 'EndDate_22',
 'EndDate_23',
 'EndDate_24',
 'EndDate_25',
 'EndDate_26',
 'EndDate_27',
 'EndDate_28',
 'EndDate_29',
 'EndDate_30']

Here I am making an list of names. It is not a dictionary because separate function takes in a list. Example is ['col1']

In [11]:
(artwork
 >> separate(X.EndDate_strip, EndDate_names,
             remove=False, convert=True,
            extra='drop', fill='right')
>> select(X.EndDate_strip,X.EndDate_0,X.EndDate_1,X.EndDate_2))

EndDate_strip EndDate_0  EndDate_1  EndDate_2
0               1918      1918        NaN        NaN
1                  0         0        NaN        NaN
2               1957      1957        NaN        NaN
3                  0         0        NaN        NaN
4               1957      1957        NaN        NaN
...              ...       ...        ...        ...
138146   0 1934 1933         0     1934.0     1933.0
138147          1943      1943        NaN        NaN
138148          1943      1943        NaN        NaN
138149          1943      1943        NaN        NaN
138150          1943      1943        NaN        NaN

[138151 rows x 4 columns]

I am applying the separate function on the artwork data set. Plugging in the list of names I made before into the correct spot (EndDate_names). Since there are 31 columns being made the first couple get cut off from the viewing. So I selected the first 3 columns which I know has data in it to see if it worked correctly. Row 138146 appears to have worked nicely. The code below is what it looks like when not selecting a couple of columns.

In [12]:
(artwork
 >> separate(X.EndDate_strip, EndDate_names,
             remove=False, convert=True,
            extra='drop', fill='right'))

Title  \
0       Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1       City of Music, National Superior Conservatory ...   
2       Villa near Vienna Project, Outside Vienna, Aus...   
3       The Manhattan Transcripts Project, New York, N...   
4       Villa, project, outside Vienna, Austria, Exter...   
...                                                   ...   
138146                                           Untitled   
138147  Plate (folio 2 verso) from Muscheln und schirm...   
138148  Plate (folio 6) from Muscheln und schirme (She...   
138149  Plate (folio 12) from Muscheln und schirme (Sh...   
138150  Plate (folio 16) from Muscheln und schirme (Sh...   

                                                   Artist  \
0                                             Otto Wagner   
1                                Christian de Portzamparc   
2                                              Emil Hoppe   
3                                         Bernard Tschumi   
4                                              Emil Hoppe   
...                                                   ...   
138146  Chesnutt Brothers Studio, Andrew Chesnutt, Lew...   
138147                                 Sophie Taeuber-Arp   
138148                                 Sophie Taeuber-Arp   
138149                                 Sophie Taeuber-Arp   
138150                                 Sophie Taeuber-Arp   

                 ConstituentID                                    ArtistBio  \
0                         6210                        (Austrian, 1841–1918)   
1                         7470                          (French, born 1944)   
2                         7605                        (Austrian, 1876–1957)   
3                         7056    (French and Swiss, born Switzerland 1944)   
4                         7605                        (Austrian, 1876–1957)   
...                        ...                                          ...   
138146  133005, 133006, 133007  (American, 1861–1934) (American, 1860–1933)   
138147                    5777                           (Swiss, 1889–1943)   
138148                    5777                           (Swiss, 1889–1943)   
138149                    5777                           (Swiss, 1889–1943)   
138150                    5777                           (Swiss, 1889–1943)   

                     Nationality          BeginDate            EndDate  \
0                     (Austrian)             (1841)             (1918)   
1                       (French)             (1944)                (0)   
2                     (Austrian)             (1876)             (1957)   
3                             ()             (1944)                (0)   
4                     (Austrian)             (1876)             (1957)   
...                          ...                ...                ...   
138146  () (American) (American)  (0) (1861) (1860)  (0) (1934) (1933)   
138147                   (Swiss)             (1889)             (1943)   
138148                   (Swiss)             (1889)             (1943)   
138149                   (Swiss)             (1889)             (1943)   
138150                   (Swiss)             (1889)             (1943)   

                  Gender     Date  \
0                 (Male)     1896   
1                 (Male)     1987   
2                 (Male)     1903   
3                 (Male)     1980   
4                 (Male)     1903   
...                  ...      ...   
138146  () (Male) (Male)  c. 1890   
138147          (Female)     1939   
138148          (Female)     1939   
138149          (Female)     1939   
138150          (Female)     1939   

                                                   Medium  ... EndDate_21  \
0           Ink and cut-and-pasted painted pages on paper  ...        NaN   
1                       Paint and colored pencil on print  ...        NaN   
2       Graphite, pen, color pencil, ink, and gouache ...  ...        Na

## Problem 2
    
**Task:** Fix the issues with `artwork.Gender`

In [13]:
artwork.Gender.unique()

array(['(Male)', '(Male) (Male)', '(Male) (Female)',
       '(Male) (Male) (Male)', '() (Female) (Male)', '(Female)',
       '(Male) (Female) (Male) (Female)', '() (Male) (Male)',
       '() (Male) (Male) (Female)', '(Male) (Male) (Female) (Female)',
       '(Male) (Male) (Male) (Male) (Male)',
       '(Male) (Female) (Female) (Male)',
       '(Male) () (Male) (Female) (Male) (Male) (Male) (Male) (Female)',
       '() (Male)', '()', '() (Male) (Male) (Male) (Male) (Male) (Male)',
       '(Male) () ()', '() (Male) (Male) (Male) (Male) (Male)',
       '(Female) (Male)',
       '() (Male) (Male) (Male) (Male) (Male) (Male) (Male)',
       '() (Male) (Male) (Male) (Female) (Male)',
       '(Male) (Male) (Male) (Male) (Female) (Male) (Male) (Male) (Male) (Male)',
       '(Male) (Male) (Male) (Male) (Female) (Male)',
       '(Male) (Female) (Male)', '(Male) (Male) () ()',
       '(Male) (Male) (Male) (Male) (Male) (Male)',
       '() (Male) (Male) (Male)',
       '(Male) (Male) (Female) (Mal

Checking the unique varaibles for the gender column. This is a complete mess as their are alot of unique variables in the same row. Their should be 31 as discussed for the EndDate.

In [14]:
artwork = (artwork
>> mutate(Gender_Stripped = X.Gender.str.replace('[()]', ''))
)

/home/vn6415dw/.pyenv/versions/anaconda3-2020.02/lib/python3.7/site-packages/dfply/base.py:82: FutureWarning: The default value of regex will change from True to False in a future version.
  **_context_kwargs(kwargs)(x)),


Removing the () around the values in the Gender column. (Male) -> Male, (Male) (Male) -> Male Male

In [15]:
artwork['Gender_Stripped'] = artwork['Gender_Stripped'].astype(str)

Making the dtype of the Gender_Stripped column (clean) into a str for separate function use

In [16]:
Gender_names = ['Gender_{0}'.format(c) for c in range(31)]
Gender_names

['Gender_0',
 'Gender_1',
 'Gender_2',
 'Gender_3',
 'Gender_4',
 'Gender_5',
 'Gender_6',
 'Gender_7',
 'Gender_8',
 'Gender_9',
 'Gender_10',
 'Gender_11',
 'Gender_12',
 'Gender_13',
 'Gender_14',
 'Gender_15',
 'Gender_16',
 'Gender_17',
 'Gender_18',
 'Gender_19',
 'Gender_20',
 'Gender_21',
 'Gender_22',
 'Gender_23',
 'Gender_24',
 'Gender_25',
 'Gender_26',
 'Gender_27',
 'Gender_28',
 'Gender_29',
 'Gender_30']

Making a list of 31 names for gender

In [17]:
(artwork
 >> separate(X.Gender_Stripped, Gender_names,
             remove=False, convert=True,
            extra='drop', fill='right')
>> select(X.Gender_Stripped,X.Gender_1,X.Gender_2,X.Gender_3))

Gender_Stripped Gender_1 Gender_2 Gender_3
0                 Male      NaN      NaN      NaN
1                 Male      NaN      NaN      NaN
2                 Male      NaN      NaN      NaN
3                 Male      NaN      NaN      NaN
4                 Male      NaN      NaN      NaN
...                ...      ...      ...      ...
138146       Male Male     Male     Male      NaN
138147          Female      NaN      NaN      NaN
138148          Female      NaN      NaN      NaN
138149          Female      NaN      NaN      NaN
138150          Female      NaN      NaN      NaN

[138151 rows x 4 columns]

I am applying the separate function on the artwork data set. Plugging in the list of gender names I made before into the correct spot (Gender_names). Since there are 31 columns being made the first couple get cut off from the viewing. So I selected the first 3 columns which I know has data in it to see if it worked correctly. Row 138146 appears to have worked since it contianed 'Male Male' in the same spot. The code below is what it looks like when not selecting a couple of columns.

In [18]:
artwork = (artwork
 >> separate(X.Gender_Stripped, Gender_names,
             remove=False, convert=True,
            extra='drop', fill='right'))

## Problem 3
    
**Task:** Fix the issues with `artwork.BeginDate` and `artwork.EndDate`.  Note that you will want to convert to `.astype('int64')` at the end of the expression.

In [20]:
artwork = (artwork
>> mutate(BeginDate_strip = X.BeginDate.str.replace('[()]', ''))
)
artwork

/home/vn6415dw/.pyenv/versions/anaconda3-2020.02/lib/python3.7/site-packages/dfply/base.py:82: FutureWarning: The default value of regex will change from True to False in a future version.
  **_context_kwargs(kwargs)(x)),


Title  \
0       Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1       City of Music, National Superior Conservatory ...   
2       Villa near Vienna Project, Outside Vienna, Aus...   
3       The Manhattan Transcripts Project, New York, N...   
4       Villa, project, outside Vienna, Austria, Exter...   
...                                                   ...   
138146                                           Untitled   
138147  Plate (folio 2 verso) from Muscheln und schirm...   
138148  Plate (folio 6) from Muscheln und schirme (She...   
138149  Plate (folio 12) from Muscheln und schirme (Sh...   
138150  Plate (folio 16) from Muscheln und schirme (Sh...   

                                                   Artist  \
0                                             Otto Wagner   
1                                Christian de Portzamparc   
2                                              Emil Hoppe   
3                                         Bernard Tschumi   
4                                              Emil Hoppe   
...                                                   ...   
138146  Chesnutt Brothers Studio, Andrew Chesnutt, Lew...   
138147                                 Sophie Taeuber-Arp   
138148                                 Sophie Taeuber-Arp   
138149                                 Sophie Taeuber-Arp   
138150                                 Sophie Taeuber-Arp   

                 ConstituentID                                    ArtistBio  \
0                         6210                        (Austrian, 1841–1918)   
1                         7470                          (French, born 1944)   
2                         7605                        (Austrian, 1876–1957)   
3                         7056    (French and Swiss, born Switzerland 1944)   
4                         7605                        (Austrian, 1876–1957)   
...                        ...                                          ...   
138146  133005, 133006, 133007  (American, 1861–1934) (American, 1860–1933)   
138147                    5777                           (Swiss, 1889–1943)   
138148                    5777                           (Swiss, 1889–1943)   
138149                    5777                           (Swiss, 1889–1943)   
138150                    5777                           (Swiss, 1889–1943)   

                     Nationality          BeginDate            EndDate  \
0                     (Austrian)             (1841)             (1918)   
1                       (French)             (1944)                (0)   
2                     (Austrian)             (1876)             (1957)   
3                             ()             (1944)                (0)   
4                     (Austrian)             (1876)             (1957)   
...                          ...                ...                ...   
138146  () (American) (American)  (0) (1861) (1860)  (0) (1934) (1933)   
138147                   (Swiss)             (1889)             (1943)   
138148                   (Swiss)             (1889)             (1943)   
138149                   (Swiss)             (1889)             (1943)   
138150                   (Swiss)             (1889)             (1943)   

                  Gender     Date  \
0                 (Male)     1896   
1                 (Male)     1987   
2                 (Male)     1903   
3                 (Male)     1980   
4                 (Male)     1903   
...                  ...      ...   
138146  () (Male) (Male)  c. 1890   
138147          (Female)     1939   
138148          (Female)     1939   
138149          (Female)     1939   
138150          (Female)     1939   

                                                   Medium  ... Gender_22  \
0           Ink and cut-and-pasted painted pages on paper  ...       NaN   
1                       Paint and colored pencil on print  ...       NaN   
2       Graphite, pen, color pencil, ink, and gouache ...  ...       NaN   

In [21]:
artwork = (artwork
>> mutate(EndDate_strip = X.EndDate.str.replace('[()]', ''))
)

/home/vn6415dw/.pyenv/versions/anaconda3-2020.02/lib/python3.7/site-packages/dfply/base.py:82: FutureWarning: The default value of regex will change from True to False in a future version.
  **_context_kwargs(kwargs)(x)),


Cleaning the BeginDate column by removing the () around the numbers

In [22]:
artwork['BeginDate_strip'] = artwork['BeginDate_strip'].astype(str)
artwork['EndDate_strip'] = artwork['EndDate_strip'].astype(str)

I already cleaned EndDate but I am putting the code down here as well. Here I am making BeginDate and EndDate dtypes to strings

In [74]:
BeginDate_names = ['BeginDate_{0}'.format(c) for c in range(31)]
EndDate_names = ['EndDate_{0}'.format(c) for c in range(31)]

Making the 31 names for BeginDate and EndDate

In [64]:
artwork = (artwork
 >>separate(X.BeginDate_strip, BeginDate_names,remove=False, convert=True,extra='drop', fill='right')
 >>separate(X.EndDate_strip, EndDate_names,remove=False, convert=True,extra='drop', fill='right'))

In [25]:
(artwork
 >> select(X.BeginDate_strip,X.BeginDate_0,X.BeginDate_1,X.BeginDate_2,
           X.EndDate_strip,X.EndDate_0,X.EndDate_1,X.EndDate_2))

BeginDate_strip BeginDate_0  BeginDate_1  BeginDate_2 EndDate_strip  \
0                 1841        1841          NaN          NaN          1918   
1                 1944        1944          NaN          NaN             0   
2                 1876        1876          NaN          NaN          1957   
3                 1944        1944          NaN          NaN             0   
4                 1876        1876          NaN          NaN          1957   
...                ...         ...          ...          ...           ...   
138146     0 1861 1860           0       1861.0       1860.0   0 1934 1933   
138147            1889        1889          NaN          NaN          1943   
138148            1889        1889          NaN          NaN          1943   
138149            1889        1889          NaN          NaN          1943   
138150            1889        1889          NaN          NaN          1943   

       EndDate_0  EndDate_1  EndDate_2  
0           1918        NaN        NaN  
1              0        NaN        NaN  
2           1957        NaN        NaN  
3              0        NaN        NaN  
4           1957        NaN        NaN  
...          ...        ...        ...  
138146         0     1934.0     1933.0  
138147      1943        NaN        NaN  
138148      1943        NaN        NaN  
138149      1943        NaN        NaN  
138150      1943        NaN        NaN  

[138151 rows x 8 columns]

Applying the two separate functions for both BeginDate and EndDate. I saved in into artwork_dates so I can do artwork_dates.columns to see if it applied correctly which is seen below.

In [26]:
list(artwork.columns)

['Title',
 'Artist',
 'ConstituentID',
 'ArtistBio',
 'Nationality',
 'BeginDate',
 'EndDate',
 'Gender',
 'Date',
 'Medium',
 'Dimensions',
 'CreditLine',
 'AccessionNumber',
 'Classification',
 'Department',
 'DateAcquired',
 'Cataloged',
 'ObjectID',
 'URL',
 'ThumbnailURL',
 'Circumference_cm',
 'Depth_cm',
 'Diameter_cm',
 'Height_cm',
 'Length_cm',
 'Weight_kg',
 'Width_cm',
 'Seat_Height_cm',
 'Duration_sec',
 'EndDate_strip',
 'Gender_Stripped',
 'Gender_0',
 'Gender_1',
 'Gender_2',
 'Gender_3',
 'Gender_4',
 'Gender_5',
 'Gender_6',
 'Gender_7',
 'Gender_8',
 'Gender_9',
 'Gender_10',
 'Gender_11',
 'Gender_12',
 'Gender_13',
 'Gender_14',
 'Gender_15',
 'Gender_16',
 'Gender_17',
 'Gender_18',
 'Gender_19',
 'Gender_20',
 'Gender_21',
 'Gender_22',
 'Gender_23',
 'Gender_24',
 'Gender_25',
 'Gender_26',
 'Gender_27',
 'Gender_28',
 'Gender_29',
 'Gender_30',
 'BeginDate_strip',
 'BeginDate_0',
 'BeginDate_1',
 'BeginDate_2',
 'BeginDate_3',
 'BeginDate_4',
 'BeginDate_5',
 '

Here I am making sure the columns applied correctly. I am selecting the orginal clean column and this the first three EndDate and BeginDates. Everything seemed to work

## Problem 4
    
**Task:** Use `value_counts` to identify some issues with `artwork.Artist` and `artwork.Nationality`.  Discuss some strategies to solve these issues.  Do your best to clean these columns up.

In [27]:
artwork = (artwork
>>mutate(ArtistStrip = X.Artist.str.replace(' ','')))
artwork

Title  \
0       Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1       City of Music, National Superior Conservatory ...   
2       Villa near Vienna Project, Outside Vienna, Aus...   
3       The Manhattan Transcripts Project, New York, N...   
4       Villa, project, outside Vienna, Austria, Exter...   
...                                                   ...   
138146                                           Untitled   
138147  Plate (folio 2 verso) from Muscheln und schirm...   
138148  Plate (folio 6) from Muscheln und schirme (She...   
138149  Plate (folio 12) from Muscheln und schirme (Sh...   
138150  Plate (folio 16) from Muscheln und schirme (Sh...   

                                                   Artist  \
0                                             Otto Wagner   
1                                Christian de Portzamparc   
2                                              Emil Hoppe   
3                                         Bernard Tschumi   
4                                              Emil Hoppe   
...                                                   ...   
138146  Chesnutt Brothers Studio, Andrew Chesnutt, Lew...   
138147                                 Sophie Taeuber-Arp   
138148                                 Sophie Taeuber-Arp   
138149                                 Sophie Taeuber-Arp   
138150                                 Sophie Taeuber-Arp   

                 ConstituentID                                    ArtistBio  \
0                         6210                        (Austrian, 1841–1918)   
1                         7470                          (French, born 1944)   
2                         7605                        (Austrian, 1876–1957)   
3                         7056    (French and Swiss, born Switzerland 1944)   
4                         7605                        (Austrian, 1876–1957)   
...                        ...                                          ...   
138146  133005, 133006, 133007  (American, 1861–1934) (American, 1860–1933)   
138147                    5777                           (Swiss, 1889–1943)   
138148                    5777                           (Swiss, 1889–1943)   
138149                    5777                           (Swiss, 1889–1943)   
138150                    5777                           (Swiss, 1889–1943)   

                     Nationality          BeginDate            EndDate  \
0                     (Austrian)             (1841)             (1918)   
1                       (French)             (1944)                (0)   
2                     (Austrian)             (1876)             (1957)   
3                             ()             (1944)                (0)   
4                     (Austrian)             (1876)             (1957)   
...                          ...                ...                ...   
138146  () (American) (American)  (0) (1861) (1860)  (0) (1934) (1933)   
138147                   (Swiss)             (1889)             (1943)   
138148                   (Swiss)             (1889)             (1943)   
138149                   (Swiss)             (1889)             (1943)   
138150                   (Swiss)             (1889)             (1943)   

                  Gender     Date  \
0                 (Male)     1896   
1                 (Male)     1987   
2                 (Male)     1903   
3                 (Male)     1980   
4                 (Male)     1903   
...                  ...      ...   
138146  () (Male) (Male)  c. 1890   
138147          (Female)     1939   
138148          (Female)     1939   
138149          (Female)     1939   
138150          (Female)     1939   

                                                   Medium  ... EndDate_22  \
0           Ink and cut-and-pasted painted pages on paper  ...        NaN   
1                       Paint and colored pencil on print  ...        NaN   
2       Graphite, pen, color pencil, ink, and gouache ...  ...        Na

Here I am taking out the spaces in the Artist names which will help me later on

In [28]:
artwork = (artwork
>>mutate(ArtistStrip = X.ArtistStrip.str.replace('-','')))
artwork

Title  \
0       Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1       City of Music, National Superior Conservatory ...   
2       Villa near Vienna Project, Outside Vienna, Aus...   
3       The Manhattan Transcripts Project, New York, N...   
4       Villa, project, outside Vienna, Austria, Exter...   
...                                                   ...   
138146                                           Untitled   
138147  Plate (folio 2 verso) from Muscheln und schirm...   
138148  Plate (folio 6) from Muscheln und schirme (She...   
138149  Plate (folio 12) from Muscheln und schirme (Sh...   
138150  Plate (folio 16) from Muscheln und schirme (Sh...   

                                                   Artist  \
0                                             Otto Wagner   
1                                Christian de Portzamparc   
2                                              Emil Hoppe   
3                                         Bernard Tschumi   
4                                              Emil Hoppe   
...                                                   ...   
138146  Chesnutt Brothers Studio, Andrew Chesnutt, Lew...   
138147                                 Sophie Taeuber-Arp   
138148                                 Sophie Taeuber-Arp   
138149                                 Sophie Taeuber-Arp   
138150                                 Sophie Taeuber-Arp   

                 ConstituentID                                    ArtistBio  \
0                         6210                        (Austrian, 1841–1918)   
1                         7470                          (French, born 1944)   
2                         7605                        (Austrian, 1876–1957)   
3                         7056    (French and Swiss, born Switzerland 1944)   
4                         7605                        (Austrian, 1876–1957)   
...                        ...                                          ...   
138146  133005, 133006, 133007  (American, 1861–1934) (American, 1860–1933)   
138147                    5777                           (Swiss, 1889–1943)   
138148                    5777                           (Swiss, 1889–1943)   
138149                    5777                           (Swiss, 1889–1943)   
138150                    5777                           (Swiss, 1889–1943)   

                     Nationality          BeginDate            EndDate  \
0                     (Austrian)             (1841)             (1918)   
1                       (French)             (1944)                (0)   
2                     (Austrian)             (1876)             (1957)   
3                             ()             (1944)                (0)   
4                     (Austrian)             (1876)             (1957)   
...                          ...                ...                ...   
138146  () (American) (American)  (0) (1861) (1860)  (0) (1934) (1933)   
138147                   (Swiss)             (1889)             (1943)   
138148                   (Swiss)             (1889)             (1943)   
138149                   (Swiss)             (1889)             (1943)   
138150                   (Swiss)             (1889)             (1943)   

                  Gender     Date  \
0                 (Male)     1896   
1                 (Male)     1987   
2                 (Male)     1903   
3                 (Male)     1980   
4                 (Male)     1903   
...                  ...      ...   
138146  () (Male) (Male)  c. 1890   
138147          (Female)     1939   
138148          (Female)     1939   
138149          (Female)     1939   
138150          (Female)     1939   

                                                   Medium  ... EndDate_22  \
0           Ink and cut-and-pasted painted pages on paper  ...        NaN   
1                       Paint and colored pencil on print  ...        NaN   
2       Graphite, pen, color pencil, ink, and gouache ...  ...        Na

Replacing the '-' with nothing for the names that have it

In [29]:
 artwork = (artwork
>>mutate(ArtistStrip = X.ArtistStrip.str.replace(',',' ')))
artwork

Title  \
0       Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1       City of Music, National Superior Conservatory ...   
2       Villa near Vienna Project, Outside Vienna, Aus...   
3       The Manhattan Transcripts Project, New York, N...   
4       Villa, project, outside Vienna, Austria, Exter...   
...                                                   ...   
138146                                           Untitled   
138147  Plate (folio 2 verso) from Muscheln und schirm...   
138148  Plate (folio 6) from Muscheln und schirme (She...   
138149  Plate (folio 12) from Muscheln und schirme (Sh...   
138150  Plate (folio 16) from Muscheln und schirme (Sh...   

                                                   Artist  \
0                                             Otto Wagner   
1                                Christian de Portzamparc   
2                                              Emil Hoppe   
3                                         Bernard Tschumi   
4                                              Emil Hoppe   
...                                                   ...   
138146  Chesnutt Brothers Studio, Andrew Chesnutt, Lew...   
138147                                 Sophie Taeuber-Arp   
138148                                 Sophie Taeuber-Arp   
138149                                 Sophie Taeuber-Arp   
138150                                 Sophie Taeuber-Arp   

                 ConstituentID                                    ArtistBio  \
0                         6210                        (Austrian, 1841–1918)   
1                         7470                          (French, born 1944)   
2                         7605                        (Austrian, 1876–1957)   
3                         7056    (French and Swiss, born Switzerland 1944)   
4                         7605                        (Austrian, 1876–1957)   
...                        ...                                          ...   
138146  133005, 133006, 133007  (American, 1861–1934) (American, 1860–1933)   
138147                    5777                           (Swiss, 1889–1943)   
138148                    5777                           (Swiss, 1889–1943)   
138149                    5777                           (Swiss, 1889–1943)   
138150                    5777                           (Swiss, 1889–1943)   

                     Nationality          BeginDate            EndDate  \
0                     (Austrian)             (1841)             (1918)   
1                       (French)             (1944)                (0)   
2                     (Austrian)             (1876)             (1957)   
3                             ()             (1944)                (0)   
4                     (Austrian)             (1876)             (1957)   
...                          ...                ...                ...   
138146  () (American) (American)  (0) (1861) (1860)  (0) (1934) (1933)   
138147                   (Swiss)             (1889)             (1943)   
138148                   (Swiss)             (1889)             (1943)   
138149                   (Swiss)             (1889)             (1943)   
138150                   (Swiss)             (1889)             (1943)   

                  Gender     Date  \
0                 (Male)     1896   
1                 (Male)     1987   
2                 (Male)     1903   
3                 (Male)     1980   
4                 (Male)     1903   
...                  ...      ...   
138146  () (Male) (Male)  c. 1890   
138147          (Female)     1939   
138148          (Female)     1939   
138149          (Female)     1939   
138150          (Female)     1939   

                                                   Medium  ... EndDate_22  \
0           Ink and cut-and-pasted painted pages on paper  ...        NaN   
1                       Paint and colored pencil on print  ...        NaN   
2       Graphite, pen, color pencil, ink, and gouache ...  ...        Na

Replacing the columns that have multiple names which contain commas with a space. This will help split the names later.

In [30]:
artwork = (artwork
>>mutate(ArtistStrip = X.ArtistStrip.str.replace('ChristiandePortzamparc','ChristianDePortzamparc'))
>>mutate(ArtistStrip = X.ArtistStrip.str.replace('Unknownphotographer','UnkownPhotographer'))
>>mutate(ArtistStrip = X.ArtistStrip.str.replace('LudwigMiesvanderRohe ','LudwigMiesVanDerRohe '))
          )
artwork

Title  \
0       Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1       City of Music, National Superior Conservatory ...   
2       Villa near Vienna Project, Outside Vienna, Aus...   
3       The Manhattan Transcripts Project, New York, N...   
4       Villa, project, outside Vienna, Austria, Exter...   
...                                                   ...   
138146                                           Untitled   
138147  Plate (folio 2 verso) from Muscheln und schirm...   
138148  Plate (folio 6) from Muscheln und schirme (She...   
138149  Plate (folio 12) from Muscheln und schirme (Sh...   
138150  Plate (folio 16) from Muscheln und schirme (Sh...   

                                                   Artist  \
0                                             Otto Wagner   
1                                Christian de Portzamparc   
2                                              Emil Hoppe   
3                                         Bernard Tschumi   
4                                              Emil Hoppe   
...                                                   ...   
138146  Chesnutt Brothers Studio, Andrew Chesnutt, Lew...   
138147                                 Sophie Taeuber-Arp   
138148                                 Sophie Taeuber-Arp   
138149                                 Sophie Taeuber-Arp   
138150                                 Sophie Taeuber-Arp   

                 ConstituentID                                    ArtistBio  \
0                         6210                        (Austrian, 1841–1918)   
1                         7470                          (French, born 1944)   
2                         7605                        (Austrian, 1876–1957)   
3                         7056    (French and Swiss, born Switzerland 1944)   
4                         7605                        (Austrian, 1876–1957)   
...                        ...                                          ...   
138146  133005, 133006, 133007  (American, 1861–1934) (American, 1860–1933)   
138147                    5777                           (Swiss, 1889–1943)   
138148                    5777                           (Swiss, 1889–1943)   
138149                    5777                           (Swiss, 1889–1943)   
138150                    5777                           (Swiss, 1889–1943)   

                     Nationality          BeginDate            EndDate  \
0                     (Austrian)             (1841)             (1918)   
1                       (French)             (1944)                (0)   
2                     (Austrian)             (1876)             (1957)   
3                             ()             (1944)                (0)   
4                     (Austrian)             (1876)             (1957)   
...                          ...                ...                ...   
138146  () (American) (American)  (0) (1861) (1860)  (0) (1934) (1933)   
138147                   (Swiss)             (1889)             (1943)   
138148                   (Swiss)             (1889)             (1943)   
138149                   (Swiss)             (1889)             (1943)   
138150                   (Swiss)             (1889)             (1943)   

                  Gender     Date  \
0                 (Male)     1896   
1                 (Male)     1987   
2                 (Male)     1903   
3                 (Male)     1980   
4                 (Male)     1903   
...                  ...      ...   
138146  () (Male) (Male)  c. 1890   
138147          (Female)     1939   
138148          (Female)     1939   
138149          (Female)     1939   
138150          (Female)     1939   

                                                   Medium  ... EndDate_22  \
0           Ink and cut-and-pasted painted pages on paper  ...        NaN   
1                       Paint and colored pencil on print  ...        NaN   
2       Graphite, pen, color pencil, ink, and gouache ...  ...        Na

I am fixing the names that could cause issues. The plan is to space the names back out using regex which will split when it sees a lowercase and uppercase letter. So i Had to fix these names so it will follow pattern

In [31]:
list(artwork.ArtistStrip.unique())

['OttoWagner',
 'ChristianDePortzamparc',
 'EmilHoppe',
 'BernardTschumi',
 'LouisI.Kahn',
 'MarcelKammerer',
 'OttoSchönthal',
 'HansPoelzig',
 'RaimundAbraham',
 'PeterEisenman RobertCole',
 'RemKoolhaas MadelonVriesendorp',
 'RogerC.Ferri',
 'MichaelGraves',
 'AldoRossi GianniBraghieri M.Bosshard',
 'LudwigMiesvanderRohe',
 'ErikGunnarAsplund',
 'ErikGunnarAsplund SigurdLewerentz',
 'PaulNelson FrantzJourdain OscarNitzchke',
 'PaulNelson OscarNitzchke FrantzJourdain',
 'StevenHoll',
 'ErichMendelsohn',
 'PaoloSoleri',
 'PaulRudolph',
 'PeterCook',
 'Diller+Scofidio ElizabethDiller RicardoScofidio',
 'ZahaHadid',
 'FumihikoMaki',
 'RemKoolhaas ZoeZenghelis EliaZenghelis MadelonVriesendorp',
 'FrankLloydWright',
 'MarioBellini',
 'VenturiandRauch RobertVenturi JohnRauch',
 'VenturiandRauch RobertVenturi JohnRauch DeniseScottBrown',
 'Venturi RauchandScottBrown RobertVenturi JohnRauch DeniseScottBrown',
 'EmilioAmbasz',
 'TadaoAndo',
 'ArataIsozaki',
 'TheovanDoesburg CornelisvanEester

Looking to see if there are anymore names that need fixing

In [32]:
Artist_names = ['Artist_{0}'.format(c) for c in range(31)]
Artist_names

['Artist_0',
 'Artist_1',
 'Artist_2',
 'Artist_3',
 'Artist_4',
 'Artist_5',
 'Artist_6',
 'Artist_7',
 'Artist_8',
 'Artist_9',
 'Artist_10',
 'Artist_11',
 'Artist_12',
 'Artist_13',
 'Artist_14',
 'Artist_15',
 'Artist_16',
 'Artist_17',
 'Artist_18',
 'Artist_19',
 'Artist_20',
 'Artist_21',
 'Artist_22',
 'Artist_23',
 'Artist_24',
 'Artist_25',
 'Artist_26',
 'Artist_27',
 'Artist_28',
 'Artist_29',
 'Artist_30']

Making 31 columns for the artists when they will be separated

In [33]:
artwork['ArtistStrip'] = artwork['ArtistStrip'].astype(str)

Converting the ArtistStrip from object to type string

In [34]:
(artwork
>>separate(X.ArtistStrip, Artist_names ,remove=False, convert=True,extra='drop', fill='right')
>>select(X.ArtistStrip,X.Artist_0,X.Artist_1,X.Artist_2))

ArtistStrip  \
0                                              OttoWagner   
1                                  ChristianDePortzamparc   
2                                               EmilHoppe   
3                                          BernardTschumi   
4                                               EmilHoppe   
...                                                   ...   
138146  ChesnuttBrothersStudio AndrewChesnutt LewisChe...   
138147                                   SophieTaeuberArp   
138148                                   SophieTaeuberArp   
138149                                   SophieTaeuberArp   
138150                                   SophieTaeuberArp   

                      Artist_0        Artist_1       Artist_2  
0                   OttoWagner             NaN            NaN  
1       ChristianDePortzamparc             NaN            NaN  
2                    EmilHoppe             NaN            NaN  
3               BernardTschumi             NaN            NaN  
4                    EmilHoppe             NaN            NaN  
...                        ...             ...            ...  
138146  ChesnuttBrothersStudio  AndrewChesnutt  LewisChesnutt  
138147        SophieTaeuberArp             NaN            NaN  
138148        SophieTaeuberArp             NaN            NaN  
138149        SophieTaeuberArp             NaN            NaN  
138150        SophieTaeuberArp             NaN            NaN  

[138151 rows x 4 columns]

In [35]:
artwork = (artwork
>>separate(X.ArtistStrip, Artist_names ,remove=False, convert=True,extra='drop', fill='right'))

Checked to see if the separate works first which is did then I saved the results into artwork

In [36]:
Artist_columns = contains('Artist_').evaluate(artwork)
Artist_columns

['Artist_0',
 'Artist_1',
 'Artist_2',
 'Artist_3',
 'Artist_4',
 'Artist_5',
 'Artist_6',
 'Artist_7',
 'Artist_8',
 'Artist_9',
 'Artist_10',
 'Artist_11',
 'Artist_12',
 'Artist_13',
 'Artist_14',
 'Artist_15',
 'Artist_16',
 'Artist_17',
 'Artist_18',
 'Artist_19',
 'Artist_20',
 'Artist_21',
 'Artist_22',
 'Artist_23',
 'Artist_24',
 'Artist_25',
 'Artist_26',
 'Artist_27',
 'Artist_28',
 'Artist_29',
 'Artist_30']

Here I am grabbing the columns that start with Artist_ and then evaluating them on the data set

In [37]:
split_names = lambda string: re.sub(r'(?<=[a-z])(?=[A-Z])', ' ', string)

In [38]:
artwork[Artist_columns] = artwork[Artist_columns].astype(str)

In [39]:

artist_names = {'{0}'.format(c):X[c].apply(split_names)
                  for c in Artist_columns}
artist_names

{'Artist_0': <dfply.base.Intention at 0x7f562a4d7c90>,
 'Artist_1': <dfply.base.Intention at 0x7f560c8642d0>,
 'Artist_2': <dfply.base.Intention at 0x7f560c864590>,
 'Artist_3': <dfply.base.Intention at 0x7f560c864850>,
 'Artist_4': <dfply.base.Intention at 0x7f560c864b10>,
 'Artist_5': <dfply.base.Intention at 0x7f560c864dd0>,
 'Artist_6': <dfply.base.Intention at 0x7f560c8650d0>,
 'Artist_7': <dfply.base.Intention at 0x7f560c865390>,
 'Artist_8': <dfply.base.Intention at 0x7f560c865650>,
 'Artist_9': <dfply.base.Intention at 0x7f560c865910>,
 'Artist_10': <dfply.base.Intention at 0x7f560c865bd0>,
 'Artist_11': <dfply.base.Intention at 0x7f560c865e90>,
 'Artist_12': <dfply.base.Intention at 0x7f560c868190>,
 'Artist_13': <dfply.base.Intention at 0x7f560c868450>,
 'Artist_14': <dfply.base.Intention at 0x7f560c868710>,
 'Artist_15': <dfply.base.Intention at 0x7f560c8689d0>,
 'Artist_16': <dfply.base.Intention at 0x7f560c868cd0>,
 'Artist_17': <dfply.base.Intention at 0x7f560c86a050>,
 '

In [40]:
artwork = (artwork
>> mutate(**artist_names))

In [41]:
(artwork
>>select(X.ArtistStrip,X.Artist_0,X.Artist_1))

ArtistStrip  \
0                                              OttoWagner   
1                                  ChristianDePortzamparc   
2                                               EmilHoppe   
3                                          BernardTschumi   
4                                               EmilHoppe   
...                                                   ...   
138146  ChesnuttBrothersStudio AndrewChesnutt LewisChe...   
138147                                   SophieTaeuberArp   
138148                                   SophieTaeuberArp   
138149                                   SophieTaeuberArp   
138150                                   SophieTaeuberArp   

                        Artist_0         Artist_1  
0                    Otto Wagner              nan  
1       Christian De Portzamparc              nan  
2                     Emil Hoppe              nan  
3                Bernard Tschumi              nan  
4                     Emil Hoppe              nan  
...                          ...              ...  
138146  Chesnutt Brothers Studio  Andrew Chesnutt  
138147        Sophie Taeuber Arp              nan  
138148        Sophie Taeuber Arp              nan  
138149        Sophie Taeuber Arp              nan  
138150        Sophie Taeuber Arp              nan  

[138151 rows x 3 columns]

The code above makes a lambda that will replace the pattern of lowercase uppercase lettesr with spaces. So HeLlo -> He Llo. I unpacked the expression using the **artist_names. Then i selected couple of columns to see if it worked and it did split correctly 

In [42]:

artwork = (artwork
>> mutate(Nationality_Stripped = X.Nationality.str.replace('[()]', ''))
)
artwork

/home/vn6415dw/.pyenv/versions/anaconda3-2020.02/lib/python3.7/site-packages/dfply/base.py:82: FutureWarning: The default value of regex will change from True to False in a future version.
  **_context_kwargs(kwargs)(x)),


Title  \
0       Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1       City of Music, National Superior Conservatory ...   
2       Villa near Vienna Project, Outside Vienna, Aus...   
3       The Manhattan Transcripts Project, New York, N...   
4       Villa, project, outside Vienna, Austria, Exter...   
...                                                   ...   
138146                                           Untitled   
138147  Plate (folio 2 verso) from Muscheln und schirm...   
138148  Plate (folio 6) from Muscheln und schirme (She...   
138149  Plate (folio 12) from Muscheln und schirme (Sh...   
138150  Plate (folio 16) from Muscheln und schirme (Sh...   

                                                   Artist  \
0                                             Otto Wagner   
1                                Christian de Portzamparc   
2                                              Emil Hoppe   
3                                         Bernard Tschumi   
4                                              Emil Hoppe   
...                                                   ...   
138146  Chesnutt Brothers Studio, Andrew Chesnutt, Lew...   
138147                                 Sophie Taeuber-Arp   
138148                                 Sophie Taeuber-Arp   
138149                                 Sophie Taeuber-Arp   
138150                                 Sophie Taeuber-Arp   

                 ConstituentID                                    ArtistBio  \
0                         6210                        (Austrian, 1841–1918)   
1                         7470                          (French, born 1944)   
2                         7605                        (Austrian, 1876–1957)   
3                         7056    (French and Swiss, born Switzerland 1944)   
4                         7605                        (Austrian, 1876–1957)   
...                        ...                                          ...   
138146  133005, 133006, 133007  (American, 1861–1934) (American, 1860–1933)   
138147                    5777                           (Swiss, 1889–1943)   
138148                    5777                           (Swiss, 1889–1943)   
138149                    5777                           (Swiss, 1889–1943)   
138150                    5777                           (Swiss, 1889–1943)   

                     Nationality          BeginDate            EndDate  \
0                     (Austrian)             (1841)             (1918)   
1                       (French)             (1944)                (0)   
2                     (Austrian)             (1876)             (1957)   
3                             ()             (1944)                (0)   
4                     (Austrian)             (1876)             (1957)   
...                          ...                ...                ...   
138146  () (American) (American)  (0) (1861) (1860)  (0) (1934) (1933)   
138147                   (Swiss)             (1889)             (1943)   
138148                   (Swiss)             (1889)             (1943)   
138149                   (Swiss)             (1889)             (1943)   
138150                   (Swiss)             (1889)             (1943)   

                  Gender     Date  \
0                 (Male)     1896   
1                 (Male)     1987   
2                 (Male)     1903   
3                 (Male)     1980   
4                 (Male)     1903   
...                  ...      ...   
138146  () (Male) (Male)  c. 1890   
138147          (Female)     1939   
138148          (Female)     1939   
138149          (Female)     1939   
138150          (Female)     1939   

                                                   Medium  ... Artist_22  \
0           Ink and cut-and-pasted painted pages on paper  ...       nan   
1                       Paint and colored pencil on print  ...       nan   
2       Graphite, pen, color pencil, ink, and gouache ...  ...       nan   

In [43]:
artwork.Nationality_Stripped.value_counts()

American                                     56963
French                                       22589
German                                        9239
British                                       5576
                                              5025
                                             ...  
French French Romanian American Romanian         1
American Russian German American American        1
American    Italian                              1
French British Portuguese French German          1
Russian Russian                                  1
Name: Nationality_Stripped, Length: 1053, dtype: int64

In [44]:
nat_names = ['Nationality_{0}'.format(c) for c in range(31)]
nat_names

['Nationality_0',
 'Nationality_1',
 'Nationality_2',
 'Nationality_3',
 'Nationality_4',
 'Nationality_5',
 'Nationality_6',
 'Nationality_7',
 'Nationality_8',
 'Nationality_9',
 'Nationality_10',
 'Nationality_11',
 'Nationality_12',
 'Nationality_13',
 'Nationality_14',
 'Nationality_15',
 'Nationality_16',
 'Nationality_17',
 'Nationality_18',
 'Nationality_19',
 'Nationality_20',
 'Nationality_21',
 'Nationality_22',
 'Nationality_23',
 'Nationality_24',
 'Nationality_25',
 'Nationality_26',
 'Nationality_27',
 'Nationality_28',
 'Nationality_29',
 'Nationality_30']

In [45]:
artwork['Nationality_Stripped'] = artwork['Nationality_Stripped'].astype(str)

In [46]:
(artwork
>>separate(X.Nationality_Stripped, nat_names ,remove=False, convert=True,extra='drop', fill='right')
>>select(X.Nationality_Stripped,X.Nationality_0,X.Nationality_1,X.Nationality_2))

Nationality_Stripped Nationality_0 Nationality_1 Nationality_2
0                  Austrian      Austrian           NaN           NaN
1                    French        French           NaN           NaN
2                  Austrian      Austrian           NaN           NaN
3                                     NaN           NaN           NaN
4                  Austrian      Austrian           NaN           NaN
...                     ...           ...           ...           ...
138146    American American           NaN      American      American
138147                Swiss         Swiss           NaN           NaN
138148                Swiss         Swiss           NaN           NaN
138149                Swiss         Swiss           NaN           NaN
138150                Swiss         Swiss           NaN           NaN

[138151 rows x 4 columns]

In [47]:
artwork = (artwork
>>separate(X.Nationality_Stripped, nat_names ,remove=False, convert=True,extra='drop', fill='right'))
artwork

Title  \
0       Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1       City of Music, National Superior Conservatory ...   
2       Villa near Vienna Project, Outside Vienna, Aus...   
3       The Manhattan Transcripts Project, New York, N...   
4       Villa, project, outside Vienna, Austria, Exter...   
...                                                   ...   
138146                                           Untitled   
138147  Plate (folio 2 verso) from Muscheln und schirm...   
138148  Plate (folio 6) from Muscheln und schirme (She...   
138149  Plate (folio 12) from Muscheln und schirme (Sh...   
138150  Plate (folio 16) from Muscheln und schirme (Sh...   

                                                   Artist  \
0                                             Otto Wagner   
1                                Christian de Portzamparc   
2                                              Emil Hoppe   
3                                         Bernard Tschumi   
4                                              Emil Hoppe   
...                                                   ...   
138146  Chesnutt Brothers Studio, Andrew Chesnutt, Lew...   
138147                                 Sophie Taeuber-Arp   
138148                                 Sophie Taeuber-Arp   
138149                                 Sophie Taeuber-Arp   
138150                                 Sophie Taeuber-Arp   

                 ConstituentID                                    ArtistBio  \
0                         6210                        (Austrian, 1841–1918)   
1                         7470                          (French, born 1944)   
2                         7605                        (Austrian, 1876–1957)   
3                         7056    (French and Swiss, born Switzerland 1944)   
4                         7605                        (Austrian, 1876–1957)   
...                        ...                                          ...   
138146  133005, 133006, 133007  (American, 1861–1934) (American, 1860–1933)   
138147                    5777                           (Swiss, 1889–1943)   
138148                    5777                           (Swiss, 1889–1943)   
138149                    5777                           (Swiss, 1889–1943)   
138150                    5777                           (Swiss, 1889–1943)   

                     Nationality          BeginDate            EndDate  \
0                     (Austrian)             (1841)             (1918)   
1                       (French)             (1944)                (0)   
2                     (Austrian)             (1876)             (1957)   
3                             ()             (1944)                (0)   
4                     (Austrian)             (1876)             (1957)   
...                          ...                ...                ...   
138146  () (American) (American)  (0) (1861) (1860)  (0) (1934) (1933)   
138147                   (Swiss)             (1889)             (1943)   
138148                   (Swiss)             (1889)             (1943)   
138149                   (Swiss)             (1889)             (1943)   
138150                   (Swiss)             (1889)             (1943)   

                  Gender     Date  \
0                 (Male)     1896   
1                 (Male)     1987   
2                 (Male)     1903   
3                 (Male)     1980   
4                 (Male)     1903   
...                  ...      ...   
138146  () (Male) (Male)  c. 1890   
138147          (Female)     1939   
138148          (Female)     1939   
138149          (Female)     1939   
138150          (Female)     1939   

                                                   Medium  ... Nationality_21  \
0           Ink and cut-and-pasted painted pages on paper  ...            NaN   
1                       Paint and colored pencil on print  ...            NaN   
2       Graphite, pen, color pencil, ink, and gouache ...  .

For the code for nationality I looked at value_counts() to help show what values there are and how many it occurs. Next I made my 31 columns and applied the separate function on the Nationality stripped column (called stripped because the () were replaced or got rid of). I selected a couple of columns to make sure it split correctly then saved it into artwork

In [50]:
Begin_columns = contains('BeginDate_').evaluate(artwork)
Begin_columns

['BeginDate_strip',
 'BeginDate_0',
 'BeginDate_1',
 'BeginDate_2',
 'BeginDate_3',
 'BeginDate_4',
 'BeginDate_5',
 'BeginDate_6',
 'BeginDate_7',
 'BeginDate_8',
 'BeginDate_9',
 'BeginDate_10',
 'BeginDate_11',
 'BeginDate_12',
 'BeginDate_13',
 'BeginDate_14',
 'BeginDate_15',
 'BeginDate_16',
 'BeginDate_17',
 'BeginDate_18',
 'BeginDate_19',
 'BeginDate_20',
 'BeginDate_21',
 'BeginDate_22',
 'BeginDate_23',
 'BeginDate_24',
 'BeginDate_25',
 'BeginDate_26',
 'BeginDate_27',
 'BeginDate_28',
 'BeginDate_29',
 'BeginDate_30']

In [81]:
End_columns = contains('EndDate_').evaluate(artwork)
End_columns

['EndDate_strip',
 'EndDate_0',
 'EndDate_1',
 'EndDate_2',
 'EndDate_3',
 'EndDate_4',
 'EndDate_5',
 'EndDate_6',
 'EndDate_7',
 'EndDate_8',
 'EndDate_9',
 'EndDate_10',
 'EndDate_11',
 'EndDate_12',
 'EndDate_13',
 'EndDate_14',
 'EndDate_15',
 'EndDate_16',
 'EndDate_17',
 'EndDate_18',
 'EndDate_19',
 'EndDate_20',
 'EndDate_21',
 'EndDate_22',
 'EndDate_23',
 'EndDate_24',
 'EndDate_25',
 'EndDate_26',
 'EndDate_27',
 'EndDate_28',
 'EndDate_29',
 'EndDate_30']

In [82]:
Begin_columns = Begin_columns[1:]
End_columns = End_columns[1:]

In [75]:
pd.to_numeric(artwork.BeginDate_0, errors = 'coerce')

0         1841.0
1         1944.0
2         1876.0
3         1944.0
4         1876.0
           ...  
138146       0.0
138147    1889.0
138148    1889.0
138149    1889.0
138150    1889.0
Name: BeginDate_0, Length: 138151, dtype: float64

In [76]:
artwork[Begin_columns] = artwork[Begin_columns].apply(pd.to_numeric, errors = 'coerce')

In [80]:
artwork.BeginDate_0.dtype

dtype('float64')

In [83]:
artwork[End_columns] = artwork[End_columns].apply(pd.to_numeric, errors = 'coerce')

In [84]:
artwork.EndDate_0.dtype

dtype('float64')

In [48]:
list(artwork.columns)

['Title',
 'Artist',
 'ConstituentID',
 'ArtistBio',
 'Nationality',
 'BeginDate',
 'EndDate',
 'Gender',
 'Date',
 'Medium',
 'Dimensions',
 'CreditLine',
 'AccessionNumber',
 'Classification',
 'Department',
 'DateAcquired',
 'Cataloged',
 'ObjectID',
 'URL',
 'ThumbnailURL',
 'Circumference_cm',
 'Depth_cm',
 'Diameter_cm',
 'Height_cm',
 'Length_cm',
 'Weight_kg',
 'Width_cm',
 'Seat_Height_cm',
 'Duration_sec',
 'EndDate_strip',
 'Gender_Stripped',
 'Gender_0',
 'Gender_1',
 'Gender_2',
 'Gender_3',
 'Gender_4',
 'Gender_5',
 'Gender_6',
 'Gender_7',
 'Gender_8',
 'Gender_9',
 'Gender_10',
 'Gender_11',
 'Gender_12',
 'Gender_13',
 'Gender_14',
 'Gender_15',
 'Gender_16',
 'Gender_17',
 'Gender_18',
 'Gender_19',
 'Gender_20',
 'Gender_21',
 'Gender_22',
 'Gender_23',
 'Gender_24',
 'Gender_25',
 'Gender_26',
 'Gender_27',
 'Gender_28',
 'Gender_29',
 'Gender_30',
 'BeginDate_strip',
 'BeginDate_0',
 'BeginDate_1',
 'BeginDate_2',
 'BeginDate_3',
 'BeginDate_4',
 'BeginDate_5',
 '